### PR TITLE
8292674: ReportJNIFatalError should print all java frames

### PR DESCRIPTION
**Pull request by dholmes-ora**
When JNI reports errors and warnings and produces a stack trace, we need to see the exact Java code that was being executed. The existing `JavaThread::print_stack` that is used is only for platform/carrier threads and doesn't show any mounted virtual thread. So we introduce a new `JavaThread::print_virtual_stack_on` method to print the mounted virtual thread's stack, and we add `JavaThread::print_jni_stack` as a utility for JNI to call that dispatches to the appropriate function based on the thread type.

The existing test `hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java` was modified to test regular and virtual threads, and to expect to see the specific stack frame of the Java native method.

Additional manual testing of the JNI `FatalError` method was done by direct fault injection in modified JDK code.

Sanity testing: Oracle tiers 1-4 (to include our -Xcheck:jni run in tier4)

Thanks,
David

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292674](https://bugs.openjdk.org/browse/JDK-8292674): ReportJNIFatalError should print all java frames

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [f4fd2893](https://git.openjdk.org/jdk/pull/11503/files/f4fd2893f958bb4690db72b765abd67863dd7095)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11503/head:pull/11503` \
`$ git checkout pull/11503`

Update a local copy of the PR: \
`$ git checkout pull/11503` \
`$ git pull https://git.openjdk.org/jdk pull/11503/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11503`

View PR using the GUI difftool: \
`$ git pr show -t 11503`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11503.diff">https://git.openjdk.org/jdk/pull/11503.diff</a>

</details>

### CODE REVIEW

**Review by pchilano — COMMENTED** [review]
Looks good to me. Some comments below.

Thanks,
Patricio

File: test/hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java [hunk 1]
```diff
@@ -24,9 +25,11 @@
 
 /**
  * @test TestPrimitiveArrayCriticalWithBadParam
- * @bug 8269697
+ * @bug 8269697 8292674
  * @summary -Xcheck:jni should catch wrong parameter passed to GetPrimitiveArrayCritical
+ * @comment Tests reporting with regular thread and virtual thread.
  * @library /test/lib
+ * @enablePreview
```
[the remark this answers is not in the document]
Comment (pchilano) [reply test/hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java:32]: Thanks for the tip @AlanBateman. I run it with -XX:-VMContinuations and verified it works.

File: src/hotspot/share/prims/jniCheck.cpp [hunk 2]
```diff
@@ -142,7 +142,7 @@ static const char * fatal_non_utf8_class_name2 = "\"";
 // When in VM state:
 static void ReportJNIWarning(JavaThread* thr, const char *msg) {
   tty->print_cr("WARNING in native method: %s", msg);
-  thr->print_stack();
+  thr->print_jni_stack();
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/prims/jniCheck.cpp:145]: That was an oversight. I've changed it now too.

File: src/hotspot/share/runtime/javaThread.cpp [hunk 3]
```diff
@@ -1715,6 +1725,48 @@ void JavaThread::print_stack_on(outputStream* st) {
...
+      }
+      cont_entry = cont_entry->parent();
+    }
+    if (f->is_java_frame()) {
+      javaVFrame* jvf = javaVFrame::cast(f);
+      java_lang_Throwable::print_stack_element(st, jvf->method(), jvf->bci());
+
+      // Print out lock information
+      if (JavaMonitorsInStackTrace) {
+        jvf->print_lock_info_on(st, count);
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/runtime/javaThread.cpp:1758]: Okay so if I want to process monitors I need to change the register map to match the one in `print_stack_on` with regards to update and process frame. Re-testing now.

File: test/hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java:32]: Added.

**Comment by dholmes-ora** [comment]
Thanks for looking at this @pchilano . I've made changes and will re-test before updating the PR. I'm retargetting this to JDK 21 now.

File: test/hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/checked/TestPrimitiveArrayCriticalWithBadParam.java:32]: GitHub hid the follow up comments from me for some reason. So I don't need to add it afterall.

**Review by pchilano — APPROVED** [review]
Looks good, thanks David!

**Comment by robehn** [comment]
Looks good, thanks!

**Comment by dholmes-ora** [comment]
Thanks for looking at this @robehn , but can you hit the approve button please :)

**Comment by robehn** [comment]
Sorry, not sure why I missed your ping.

**Pull request closed without merging**